src/utils/SpokeUtils.ts

```diff
@@ -0,0 +1,142 @@
+  // Set the initial high block to the most recent block number or the initial high block, whichever is smaller.
+  initHigh = Math.min(initHigh, mostRecentBlockNumber);
+
+  // Define a mapping of block numbers to deposit IDs. This is used to cache the deposit ID at a block number
+  // so we don't need to make an eth_call request to get the deposit ID at a block number more than once.
+  const queriedIds: Record<number, number> = {};
+
+  // Define a llambda function to get the deposit ID at a block number. This function will first check the
+  // queriedIds cache to see if the deposit ID at the block number has already been queried. If not, it will
+  // make an eth_call request to get the deposit ID at the block number. It will then cache the deposit ID
+  // in the queriedIds cache.
+  const _getDepositIdAtBlock = async (blockNumber: number): Promise<number> => {
+    if (queriedIds[blockNumber] === undefined) {
+      const resultId = await getDepositIdAtBlock(spokePool, blockNumber);
+      if (!Number.isInteger(resultId)) {
+        throw new Error("Invalid deposit count");
+      }
+      queriedIds[blockNumber] = resultId;
+    }
+    return queriedIds[blockNumber];
+  };
+
+  // Sanity check to ensure that init Low is greater than or equal to zero.
+  if (initLow < deploymentBlock) {
+    throw new Error("Binary search failed because low must be >= 0");
+  }
+
+  // Sanity check to ensure that initHigh is greater than or equal to initLow.
+  if (initLow > initHigh) {
+    throw new Error("Binary search failed because low > high");
+  }
+
+  // Sanity check to ensure that maxSearches is greater than zero.
+  if (maxSearches <= 0) {
+    throw new Error("maxSearches must be > 0");
+  }
+
+  // Sanity check to ensure that deploymentBlock is greater than or equal to zero.
+  if (deploymentBlock < 0) {
+    throw new Error("deploymentBlock must be >= 0");
+  }
+
+  // If the deposit ID at the initial high block is less than the target deposit ID, then we know that
+  // the target deposit ID must be greater than the initial high block, so we can throw an error.
+  if ((await _getDepositIdAtBlock(initHigh)) <= targetDepositId) {
+    // initLow   = 5: Deposits Num: 10
+    //                                     // targetId = 11  <- fail (triggers this error)          // 10 <= 11
+    //                                     // targetId = 10  <- fail (triggers this error)          // 10 <= 10
+    //                                     // targetId = 09  <- pass (does not trigger this error)  // 10 <= 09
+    throw new Error("Failed to find deposit ID: Target Id is greater than the initial high block");
+  }
+
+  // If the deposit ID at the initial low block is greater than the target deposit ID, then we know that
+  // the target deposit ID must be less than the initial low block, so we can throw an error.
+  if ((await _getDepositIdAtBlock(Math.max(deploymentBlock, initLow - 1))) > targetDepositId) {
```

Comment:
Consider combining this with the call to `getBlockNumber()` (at the top) so they can be executed in parallel?